File: apps/worker/src/lib/github.ts
import { config } from '@avail/shared/config';
import { createLogger } from '@avail/shared/logger';

const log = createLogger('github');

export class GitHubError extends Error {
  constructor(
    message: string,
    public status: number,
    public body?: unknown
  ) {
    super(message);
  }
}

export interface GhUser {
  login: string;
  name: string | null;
  email: string | null;
  avatar_url: string;
}

export interface GhRepo {
  id: number;
  name: string;
  full_name: string;
  private: boolean;
  html_url: string;
  description: string | null;
  default_branch: string;
  updated_at: string;
  pushed_at: string;
  owner: { login: string; avatar_url: string };
  permissions?: { admin: boolean; push: boolean; pull: boolean };
}

async function request<T>(
  token: string,
  endpoint: string,
  init: RequestInit = {}
): Promise<T> {
  const url = endpoint.startsWith('http')
    ? endpoint
    : `${config.github.apiUrl}${endpoint}`;

  const response = await fetch(url, {
    ...init,
    headers: {
      Accept: 'application/vnd.github+json',
      Authorization: `Bearer ${token}`,
      'X-GitHub-Api-Version': '2022-11-28',
      'User-Agent': 'avail-deploy',
      ...(init.body ? { 'Content-Type': 'application/json' } : {}),
      ...(init.headers as Record<string, string> | undefined),
    },
  });

  if (response.status === 204) return undefined as T;

  const text = await response.text();
  let body: unknown = text;
  try {
    body = text ? JSON.parse(text) : null;
  } catch {
    /* keep raw text */
  }

  if (!response.ok) {
    const message =
      (body as { message?: string })?.message ?? `GitHub request failed`;
    throw new GitHubError(
      `${message} (${response.status} ${endpoint})`,
      response.status,
      body
    );
  }
  return body as T;
}

export const github = {
  async user(token: string): Promise<GhUser> {
    return request<GhUser>(token, '/user');
  },

  /** Primary verified email, which is what the domain allow-list checks. */
  async primaryEmail(token: string): Promise<string | null> {
    try {
      const emails = await request<
        { email: string; primary: boolean; verified: boolean }[]
      >(token, '/user/emails');
      const primary = emails.find((e) => e.primary && e.verified);
      return primary?.email ?? emails.find((e) => e.verified)?.email ?? null;
    } catch (err) {
      log.warn('Could not read user emails:', (err as Error).message);
      return null;
    }
  },

  async repos(token: string, page = 1, perPage = 100): Promise<GhRepo[]> {
    return request<GhRepo[]>(
      token,
      `/user/repos?per_page=${perPage}&page=${page}&sort=pushed&affiliation=owner,collaborator,organization_member`
    );
  },

  async repo(token: string, fullName: string): Promise<GhRepo> {
    return request<GhRepo>(token, `/repos/${fullName}`);
  },

  async branches(token: string, fullName: string) {
    return request<{ name: string; commit: { sha: string } }[]>(
      token,
      `/repos/${fullName}/branches?per_page=100`
    );
  },

  /** Latest commit on a branch, or null when the branch does not exist. */
  async branchHead(
    token: string,
    fullName: string,
    branch: string
  ): Promise<{ sha: string; message: string; author: string; url: string } | null> {
    try {
      const data = await request<{
        sha: string;
        html_url: string;
        commit: { message: string; author: { name: string; email: string } };
      }>(token, `/repos/${fullName}/commits/${encodeURIComponent(branch)}`);
      return {
        sha: data.sha,
        message: data.commit.message.split('\n')[0],
        author: `${data.commit.author.name} <${data.commit.author.email}>`,
        url: data.html_url,
      };
    } catch (err) {
      if (err instanceof GitHubError && err.status === 404) return null;
      throw err;
    }
  },

  async createWebhook(
    token: string,
    fullName: string,
    url: string,
    secret: string
  ) {
    return request<{ id: number }>(token, `/repos/${fullName}/hooks`, {
      method: 'POST',
      body: JSON.stringify({
        name: 'web',
        active: true,
        events: ['push', 'pull_request'],
        config: { url, content_type: 'json', secret, insecure_ssl: '0' },
      }),
    });
  },

  async listWebhooks(token: string, fullName: string) {
    return request<{ id: number; config: { url?: string } }[]>(
      token,
      `/repos/${fullName}/hooks`
    );
  },

  async deleteWebhook(token: string, fullName: string, hookId: number) {
    return request<void>(token, `/repos/${fullName}/hooks/${hookId}`, {
      method: 'DELETE',
    });
  },

  /** Reports deployment status back onto the commit in GitHub. */
  async createCommitStatus(
    token: string,
    fullName: string,
    sha: string,
    status: {
      state: 'pending' | 'success' | 'failure' | 'error';
      target_url?: string;
      description?: string;
      context?: string;
    }
  ) {
    return request(token, `/repos/${fullName}/statuses/${sha}`, {
      method: 'POST',
      body: JSON.stringify({ context: 'avail/deploy', ...status }),
    });
  },

  async commentOnPullRequest(
    token: string,
    fullName: string,
    prNumber: number,
    body: string
  ) {
    return request(token, `/repos/${fullName}/issues/${prNumber}/comments`, {
      method: 'POST',
      body: JSON.stringify({ body }),
    });
  },

  /** Exchanges an OAuth callback code for an access token. */
  async exchangeOAuthCode(code: string, redirectUri: string): Promise<string> {
    if (!config.github.clientId || !config.github.clientSecret) {
      throw new GitHubError('GitHub OAuth is not configured', 500);
    }
    const response = await fetch('https://github.com/login/oauth/access_token', {
      method: 'POST',
      headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
      body: JSON.stringify({
        client_id: config.github.clientId,
        client_secret: config.github.clientSecret,
        code,
        redirect_uri: redirectUri,
      }),
    });
    const data = (await response.json()) as {
      access_token?: string;
      error_description?: string;
    };
    if (!data.access_token) {
      throw new GitHubError(
        data.error_description ?? 'OAuth exchange failed',
        400,
        data
      );
    }
    return data.access_token;
  },

  oauthAuthorizeUrl(redirectUri: string, state: string): string {
    const params = new URLSearchParams({
      client_id: config.github.clientId ?? '',
      redirect_uri: redirectUri,
      scope: 'repo read:user user:email admin:repo_hook',
      state,
    });
    return `https://github.com/login/oauth/authorize?${params}`;
  },
};

File: rust-api/src/github.rs
//! Mirrors `apps/api/src/lib/github.ts`.

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

use crate::config::GithubConfig;

#[derive(Debug)]
pub struct GitHubError {
    pub message: String,
    pub status: u16,
}

impl std::fmt::Display for GitHubError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}
impl std::error::Error for GitHubError {}

#[derive(Debug, Clone, Deserialize)]
pub struct GhUser {
    pub login: String,
    pub name: Option<String>,
    pub email: Option<String>,
    pub avatar_url: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GhOwner {
    pub login: String,
    pub avatar_url: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct GhPermissions {
    #[serde(default)]
    pub admin: bool,
}

#[derive(Debug, Clone, Deserialize)]
pub struct GhRepo {
    pub id: i64,
    pub name: String,
    pub full_name: String,
    pub private: bool,
    pub html_url: String,
    pub description: Option<String>,
    pub default_branch: String,
    pub pushed_at: String,
    pub owner: GhOwner,
    pub permissions: Option<GhPermissions>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct GhBranch {
    pub name: String,
    pub commit: GhBranchCommit,
}
#[derive(Debug, Clone, Deserialize)]
pub struct GhBranchCommit {
    pub sha: String,
}

#[derive(Debug, Clone)]
pub struct BranchHead {
    pub sha: String,
    pub message: String,
    pub author: String,
    pub url: String,
}

fn client() -> reqwest::Client {
    reqwest::Client::builder().user_agent("avail-deploy").build().expect("build reqwest client")
}

async fn request(
    api_url: &str,
    token: &str,
    method: reqwest::Method,
    endpoint: &str,
    body: Option<Value>,
) -> Result<Value, GitHubError> {
    let url = if endpoint.starts_with("http") { endpoint.to_string() } else { format!("{api_url}{endpoint}") };
    let mut req = client()
        .request(method, &url)
        .header("Accept", "application/vnd.github+json")
        .header("Authorization", format!("Bearer {token}"))
        .header("X-GitHub-Api-Version", "2022-11-28");
    if let Some(body) = &body {
        req = req.json(body);
    }

    let response = req.send().await.map_err(|e| GitHubError { message: e.to_string(), status: 0 })?;
    let status = response.status();
    if status.as_u16() == 204 {
        return Ok(Value::Null);
    }
    let text = response.text().await.unwrap_or_default();
    let parsed: Value = serde_json::from_str(&text).unwrap_or(Value::String(text));

    if !status.is_success() {
        let message = parsed.get("message").and_then(|m| m.as_str()).unwrap_or("GitHub request failed");
        return Err(GitHubError { message: format!("{message} ({} {endpoint})", status.as_u16()), status: status.as_u16() });
    }
    Ok(parsed)
}

async fn get(api_url: &str, token: &str, endpoint: &str) -> Result<Value, GitHubError> {
    request(api_url, token, reqwest::Method::GET, endpoint, None).await
}

pub async fn user(api_url: &str, token: &str) -> Result<GhUser, GitHubError> {
    let v = get(api_url, token, "/user").await?;
    serde_json::from_value(v).map_err(|e| GitHubError { message: e.to_string(), status: 0 })
}

/// Primary verified email, which is what the domain allow-list checks.
pub async fn primary_email(api_url: &str, token: &str) -> Option<String> {
    #[derive(Deserialize)]
    struct Entry {
        email: String,
        primary: bool,
        verified: bool,
    }
    let v = get(api_url, token, "/user/emails").await.ok()?;
    let emails: Vec<Entry> = serde_json::from_value(v).ok()?;
    emails
        .iter()
        .find(|e| e.primary && e.verified)
        .or_else(|| emails.iter().find(|e| e.verified))
        .map(|e| e.email.clone())
}

pub async fn repos(api_url: &str, token: &str, page: i64) -> Result<Vec<GhRepo>, GitHubError> {
    let endpoint = format!(
        "/user/repos?per_page=100&page={page}&sort=pushed&affiliation=owner,collaborator,organization_member"
    );
    let v = get(api_url, token, &endpoint).await?;
    serde_json::from_value(v).map_err(|e| GitHubError { message: e.to_string(), status: 0 })
}

pub async fn repo(api_url: &str, token: &str, full_name: &str) -> Result<GhRepo, GitHubError> {
    let v = get(api_url, token, &format!("/repos/{full_name}")).await?;
    serde_json::from_value(v).map_err(|e| GitHubError { message: e.to_string(), status: 0 })
}

pub async fn branches(api_url: &str, token: &str, full_name: &str) -> Result<Vec<GhBranch>, GitHubError> {
    let v = get(api_url, token, &format!("/repos/{full_name}/branches?per_page=100")).await?;
    serde_json::from_value(v).map_err(|e| GitHubError { message: e.to_string(), status: 0 })
}

/// Latest commit on a branch, or `None` when the branch does not exist.
pub async fn branch_head(api_url: &str, token: &str, full_name: &str, branch: &str) -> Option<BranchHead> {
    let encoded = urlencoding::encode(branch);
    let v = get(api_url, token, &format!("/repos/{full_name}/commits/{encoded}")).await.ok()?;
    let sha = v.get("sha")?.as_str()?.to_string();
    let html_url = v.get("html_url")?.as_str()?.to_string();
    let commit = v.get("commit")?;
    let message = commit.get("message")?.as_str()?.split('\n').next().unwrap_or("").to_string();
    let author = commit.get("author")?;
    let name = author.get("name").and_then(|n| n.as_str()).unwrap_or("");
    let email = author.get("email").and_then(|n| n.as_str()).unwrap_or("");
    Some(BranchHead { sha, message, author: format!("{name} <{email}>"), url: html_url })
}

pub async fn create_webhook(api_url: &str, token: &str, full_name: &str, url: &str, secret: &str) -> Result<i64, GitHubError> {
    let body = json!({
        "name": "web",
        "active": true,
        "events": ["push", "pull_request"],
        "config": { "url": url, "content_type": "json", "secret": secret, "insecure_ssl": "0" },
    });
    let v = request(api_url, token, reqwest::Method::POST, &format!("/repos/{full_name}/hooks"), Some(body)).await?;
    Ok(v.get("id").and_then(|i| i.as_i64()).unwrap_or(0))
}

pub async fn list_webhooks(api_url: &str, token: &str, full_name: &str) -> Result<Vec<(i64, Option<String>)>, GitHubError> {
    let v = get(api_url, token, &format!("/repos/{full_name}/hooks")).await?;
    let arr = v.as_array().cloned().unwrap_or_default();
    Ok(arr
        .into_iter()
        .map(|hook| {
            let id = hook.get("id").and_then(|i| i.as_i64()).unwrap_or(0);
            let url = hook.get("config").and_then(|c| c.get("url")).and_then(|u| u.as_str()).map(str::to_string);
            (id, url)
        })
        .collect())
}

pub async fn delete_webhook(api_url: &str, token: &str, full_name: &str, hook_id: i64) -> Result<(), GitHubError> {
    request(api_url, token, reqwest::Method::DELETE, &format!("/repos/{full_name}/hooks/{hook_id}"), None).await?;
    Ok(())
}

/// Reports deployment status back onto the commit in GitHub. Unused in this
/// binary today — that's `apps/worker`'s job, using its own copy of
/// `lib/github.ts`, once a build actually finishes. Ported for parity.
#[allow(dead_code)]
pub async fn create_commit_status(
    api_url: &str,
    token: &str,
    full_name: &str,
    sha: &str,
    state: &str,
    target_url: &str,
    description: &str,
    context: &str,
) -> Result<(), GitHubError> {
    let body = json!({
        "context": context,
        "state": state,
        "target_url": target_url,
        "description": description,
    });
    request(api_url, token, reqwest::Method::POST, &format!("/repos/{full_name}/statuses/{sha}"), Some(body)).await?;
    Ok(())
}

/// Also unused here for the same reason as `create_commit_status`.
#[allow(dead_code)]
pub async fn comment_on_pull_request(api_url: &str, token: &str, full_name: &str, pr_number: i64, body_text: &str) -> Result<(), GitHubError> {
    let body = json!({ "body": body_text });
    request(api_url, token, reqwest::Method::POST, &format!("/repos/{full_name}/issues/{pr_number}/comments"), Some(body)).await?;
    Ok(())
}

/// Exchanges an OAuth callback code for an access token.
pub async fn exchange_oauth_code(cfg: &GithubConfig, code: &str, redirect_uri: &str) -> Result<String, GitHubError> {
    let (Some(client_id), Some(client_secret)) = (&cfg.client_id, &cfg.client_secret) else {
        return Err(GitHubError { message: "GitHub OAuth is not configured".into(), status: 500 });
    };
    let response = client()
        .post("https://github.com/login/oauth/access_token")
        .header("Accept", "application/json")
        .json(&json!({
            "client_id": client_id,
            "client_secret": client_secret,
            "code": code,
            "redirect_uri": redirect_uri,
        }))
        .send()
        .await
        .map_err(|e| GitHubError { message: e.to_string(), status: 0 })?;
    let body: Value = response.json().await.unwrap_or(Value::Null);
    match body.get("access_token").and_then(|t| t.as_str()) {
        Some(token) => Ok(token.to_string()),
        None => {
            let desc = body.get("error_description").and_then(|d| d.as_str()).unwrap_or("OAuth exchange failed");
            Err(GitHubError { message: desc.to_string(), status: 400 })
        }
    }
}

pub fn oauth_authorize_url(cfg: &GithubConfig, redirect_uri: &str, state: &str) -> String {
    let client_id = cfg.client_id.as_deref().unwrap_or("");
    format!(
        "https://github.com/login/oauth/authorize?client_id={}&redirect_uri={}&scope={}&state={}",
        urlencoding::encode(client_id),
        urlencoding::encode(redirect_uri),
        urlencoding::encode("repo read:user user:email admin:repo_hook"),
        urlencoding::encode(state),
    )
}
